haskell/lint.bzl

```diff
@@ -0,0 +1,170 @@
+"""Lint Haskell sources using hlint."""
+
+HaskellHLintInfo = provider(
+    doc = "Provider that collects files produced by hlint",
+    fields = {
+        "outputs": "depset of hlint log files.",
+    },
+)
+
+def _collect_hlint_logs(deps):
+    return depset(transitive = [
+        dep[HaskellHLintInfo].outputs
+        for dep in deps
+        if HaskellHLintInfo in dep
+    ])
+
+def _haskell_lint_aspect_impl(target, ctx):
+    hlint_toolchain = ctx.toolchains["@io_tweag_rules_haskell//haskell/lint:toolchain"]
+    hint = hlint_toolchain.hint
+
+    inputFiles = []
+    inputPaths = []
+    if hasattr(ctx.rule.attr, "srcs"):
+        for src in ctx.rule.attr.srcs:
+            for f in src.files:
+                # We want to only do native Haskell source files, which
+                # seems to involve ignoring these generated paths
+                # (the f.is_source almost always returns True)
+                if all([
```

Comment:
Wouldn't [f.is_source](https://docs.bazel.build/versions/master/skylark/lib/File.html#is_source) filter the files as we want here?

Reply:
Good point. When this was first implemented `f.is_source` was insufficient, and these patterns also filter out external repositories. Though, it seems an approach similar to [`coverage_source_patterns`](https://github.com/tweag/rules_haskell/blob/af0644b9ab8e2bbf6b6921ffe275181cf0fe8ec2/haskell/haskell.bzl#L163-L171) or [`from_source`](https://github.com/tweag/rules_haskell/blob/af0644b9ab8e2bbf6b6921ffe275181cf0fe8ec2/haskell/repl.bzl#L369-L380) would be the cleaner solution here. I'll give it a shot.

Reply:
I've changed the code to use `f.is_source` and to allow users to control which targets to apply `hlint` to. PTAL